test/multiple-user-principal-types.test.js

```diff
@@ -0,0 +1,978 @@
+
+            done();
+          });
+        });
+      });
+    });
+
+    it('Login a user using a custom createAccessToken with options',
+    function(done) {
+      var createToken = User.prototype.createAccessToken; // Save the original method
+      // Override createAccessToken
+      User.prototype.createAccessToken = function(ttl, options, cb) {
+        // Reduce the ttl by half for testing purpose
+        this.accessTokens.create({ttl: ttl / 2, scopes: options.scope}, cb);
+      };
+      User.login(validCredentialsWithTTLAndScope, function(err, accessToken) {
+        assertGoodToken(accessToken, validCredentialsUser);
+        assert.equal(accessToken.ttl, 1800);
+        assert.equal(accessToken.scopes, 'all');
+
+        User.findById(accessToken.userId, function(err, user) {
+          user.createAccessToken(120, {scope: 'default'}, function(err, accessToken) {
+            assertGoodToken(accessToken, validCredentialsUser);
+            assert.equal(accessToken.ttl, 60);
+            assert.equal(accessToken.scopes, 'default');
+            // Restore create access token
+            User.prototype.createAccessToken = createToken;
+
+            done();
+          });
+        });
+      });
+    });
+
+    it('Login should only allow correct credentials', function(done) {
+      User.login(invalidCredentials, function(err, accessToken) {
+        assert(err);
+        assert.equal(err.code, 'LOGIN_FAILED');
+        assert(!accessToken);
+
+        done();
+      });
+    });
+
+    it('Login should only allow correct credentials - promise variant', function(done) {
+      User.login(invalidCredentials)
+      .then(function(accessToken) {
+        assert(!accessToken);
+
+        done();
+      })
+      .catch(function(err) {
+        assert(err);
+        assert.equal(err.code, 'LOGIN_FAILED');
+
+        done();
+      });
+    });
+
+    it('Login a user providing incomplete credentials', function(done) {
+      User.login(incompleteCredentials, function(err, accessToken) {
+        assert(err);
+        assert.equal(err.code, 'USERNAME_EMAIL_REQUIRED');
+
+        done();
+      });
+    });
+
+    it('Login a user providing incomplete credentials - promise variant', function(done) {
+      User.login(incompleteCredentials)
+      .then(function(accessToken) {
+        assert(!accessToken);
+
+        done();
+      })
+      .catch(function(err) {
+        assert(err);
+        assert.equal(err.code, 'USERNAME_EMAIL_REQUIRED');
+
+        done();
+      });
+    });
+  });
+
+  function assertGoodToken(accessToken, user) {
+    if (accessToken instanceof AccessToken) {
+      accessToken = accessToken.toJSON();
+    }
+    expect(accessToken).to.have.property('userId', user.id);
+    expect(accessToken).to.have.property('principalType', user.constructor.definition.name);
+    assert(accessToken.id);
+    assert.equal(accessToken.id.length, 64);
+  }
+
+  describe('User.login requiring realm', function() {
```

Comment:
Is the code handling realms related to polymorphic User relation at all? I think this whole test suite can be removed from this file.

Thoughts?